server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkSchemaUtil.java

```diff
@@ -0,0 +1,165 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Utility to compute schema for all sinks in a streaming ingestion task, used in {@link StreamAppenderator}.
+ */
+class SinkSchemaUtil
+{
+  /**
+   * Compute {@link SegmentSchemas} for the sinks.
+   */
+  @VisibleForTesting
+  static Optional<SegmentSchemas> computeAbsoluteSchema(
+      Map<SegmentId, Pair<RowSignature, Integer>> sinkSchemaMap
+  )
+  {
+    List<SegmentSchema> sinkSchemas = new ArrayList<>();
+
+    for (Map.Entry<SegmentId, Pair<RowSignature, Integer>> entry : sinkSchemaMap.entrySet()) {
+      SegmentId segmentId = entry.getKey();
+      RowSignature sinkSignature = entry.getValue().lhs;
+
+      List<String> newColumns = new ArrayList<>();
+
+      Map<String, ColumnType> columnMapping = new HashMap<>();
+
+      // new Sink
+      for (String column : sinkSignature.getColumnNames()) {
+        newColumns.add(column);
+        sinkSignature.getColumnType(column).ifPresent(type -> columnMapping.put(column, type));
+      }
+
+      Integer numRows = entry.getValue().rhs;
+      if (newColumns.size() > 0) {
+        SegmentSchema segmentSchema =
+            new SegmentSchema(
+                segmentId.getDataSource(),
+                segmentId.toString(),
+                false,
+                numRows,
+                newColumns,
+                Collections.emptyList(),
+                columnMapping
+            );
+        sinkSchemas.add(segmentSchema);
+      }
+    }
+
+    return Optional.ofNullable(sinkSchemas.isEmpty() ? null : new SegmentSchemas(sinkSchemas));
+  }
+
+  /**
+   * Compute schema change for the sinks.
+   */
+  @VisibleForTesting
+  static Optional<SegmentSchemas> computeSchemaChange(
+      Map<SegmentId, Pair<RowSignature, Integer>> previousSinkSignatureMap,
+      Map<SegmentId, Pair<RowSignature, Integer>> currentSinkSignatureMap
+  )
+  {
+    List<SegmentSchema> sinkSchemas = new ArrayList<>();
+
+    for (Map.Entry<SegmentId, Pair<RowSignature, Integer>> entry : currentSinkSignatureMap.entrySet()) {
+      SegmentId segmentId = entry.getKey();
+      RowSignature sinkSignature = entry.getValue().lhs;
+
+      Integer numRows = entry.getValue().rhs;
+
+      List<String> newColumns = new ArrayList<>();
+      List<String> updatedColumns = new ArrayList<>();
+      Map<String, ColumnType> columnMapping = new HashMap<>();
+
+      // whether there are any changes to be published
+      boolean publish = false;
+      // if the resultant schema is delta
+      boolean delta = false;
+
+      if (!previousSinkSignatureMap.containsKey(segmentId)) {
+        // new Sink
+        for (String column : sinkSignature.getColumnNames()) {
+          newColumns.add(column);
+          sinkSignature.getColumnType(column).ifPresent(type -> columnMapping.put(column, type));
+        }
+        if (newColumns.size() > 0 || numRows > 0) {
```

Comment:
```suggestion
        if (newColumns.size() > 0) {
```
I think we should only publish when we have newCols.